pkg/lib/dialogs.tsx

```diff
@@ -153,41 +158,28 @@ export const useDialogs = () => {
 };
 
 export const WithDialogs = ({ children } : { children: React.ReactNode }) => {
-    const is_open = useRef(false); // synchronous
-    const resolveRef = useRef<((x: unknown) => void) | null>(null);
-    const rejectRef = useRef<((x: unknown) => void) | null>(null);
     const [dialog, setDialog] = useState<React.ReactNode>(null);
+    const shown = useRef<"show" | "run" | "close">("close");
+
+    function transition(expected: string, to: typeof shown.current, arg: React.ReactNode = null) {
+        if (shown.current !== expected)
+            throw new Error(`Dialogs.${to}(${JSON.stringify(arg)}) called, but that's only valid` +
+                            `after .${expected}(), current dialog is ${JSON.stringify(dialog)}.`);
+        shown.current = to;
+        setDialog(arg);
+    }
 
     const Dialogs: Dialogs = {
-        show: component => {
-            if (component && is_open.current)
-                console.error("Dialogs.show() called for",
-                              JSON.stringify(component),
-                              "while a dialog is already open:",
-                              JSON.stringify(dialog));
-            is_open.current = !!component;
-            setDialog(component);
-            return new Promise((resolve, reject) => {
-                resolveRef.current = resolve;
-                rejectRef.current = reject;
-            });
-        },
-        close: (args) => {
-            is_open.current = false;
-            setDialog(null);
-            if (resolveRef.current !== null) {
-                resolveRef.current(args);
-                resolveRef.current = null;
-                rejectRef.current = null;
-            }
-        },
-        reject: (err) => {
-            is_open.current = false;
-            setDialog(null);
-            if (rejectRef.current !== null) {
-                rejectRef.current(err);
-                resolveRef.current = null;
-                rejectRef.current = null;
+        show: (component: React.ReactNode) => transition("close", "show", component),
+        close: () => transition("show", "close"),
+        run: async (component, properties) => {
+            try {
+                return await new Promise((resolve, reject) => {
```

Comment:
Unnecessary `async/await` pair here, you can just make the function sync and return the promise directly.